### PR TITLE
support passing db implementation to merkle tree

### PR DESCRIPTION
**Pull request by ilovers**

### CODE REVIEW

**Review by xgaozoyoe — APPROVED** [review]

File: src/host/mongomerkle.rs [hunk 1]
```diff
@@ -197,27 +126,37 @@ impl<const DEPTH: usize> MongoMerkle<DEPTH> {
...
+                    Ok(Some(_)) => {}
+                    _ => {
+                        check = false;
+                        new_records.push(record);
+                    }
+                }
+            } else {
+                new_records.push(record);
+            }
+        }
```
Comment (yymone) [inline src/host/mongomerkle.rs:152]: This batch_update_records logic changed.
Originally it just try insert the input records to db (if exist, skip. If not exist, insert)
Now the input records must be in one leaf path. (Must be the parend nodes of a leaf)
Currently we only use it for this scenario so if we do this change, better change the name to something like batch_update_records_in_one_leaf_path, etc

File: src/host/mongomerkle.rs [hunk 2]
```diff
@@ -395,11 +334,11 @@ impl<const DEPTH: usize> MerkleTree<[u8; 32], DEPTH> for MongoMerkle<DEPTH> {
     type Root = [u8; 32];
     type Node = MerkleRecord;
 
-    fn construct(addr: Self::Id, root: Self::Root) -> Self {
+    fn construct(addr: Self::Id, root: Self::Root, db: Option<Rc<RefCell<dyn TreeDB>>>) -> Self {
```
Comment (yymone) [inline src/host/mongomerkle.rs:337]: Wrong box, should be Option<Rc<RefCell<dyn TreeDB>>>.
Compile error when run cargo test.

File: src/host/mongomerkle.rs [hunk 3]
```diff
@@ -615,10 +557,11 @@ mod tests {
         ];
 
         // 1
-        let mut mt = MongoMerkle::<DEPTH>::construct(TEST_ADDR, DEFAULT_HASH_VEC[DEPTH].clone());
-        let dbname: String = MongoMerkle::<DEPTH>::get_db_name();
-        let cname = mt.get_collection_name();
-        let collection = get_collection::<MerkleRecord>(dbname, cname).unwrap();
+        let mut mt =
+            MongoMerkle::<DEPTH>::construct(TEST_ADDR, DEFAULT_HASH_VEC[DEPTH].clone(), None);
+        let cname = get_collection_name(MONGODB_DATA_NAME_PREFIX.to_string(), test_addr);
```
Comment (yymone) [inline src/host/mongomerkle.rs:562]: lower test_addr not defined.
Compile error when run cargo test.

File: src/host/mongomerkle.rs [hunk 4]
```diff
@@ -573,10 +514,11 @@ mod tests {
...
```
Comment (yymone) [inline src/host/mongomerkle.rs:519]: lower test_addr not defined.
Compile error when run cargo test.

File: src/host/mongomerkle.rs (hunk 1 above)
Comment (ilovers) [reply src/host/mongomerkle.rs:153]: rename batch_update_records to update_leaf_path_records

**Review by yymone — APPROVED** [review]

**Pull request merged**